tools/sense_studio/training.py

```diff
@@ -40,68 +33,88 @@ def train_model():
 
 
 @training_bp.route('/cancel-training', methods=['POST'])
 def cancel_training():
     data = request.form
     project = data['project']
     path = data['path']
-    global PROCESS
-    if PROCESS:
-        PROCESS.terminate()
-        PROCESS = None
-        log = "Training Cancelled."
+    output_folder = data['output_folder']
+    global train_process
+    if train_process:
+        train_process.terminate()
+        train_process = None
+        message = "Training Cancelled."
+    return render_template('training.html', project=project, path=path, models=utils.BACKBONE_MODELS,
+                           is_disabled=False, message=message, output_folder=output_folder)
+
+
+@socketio.on('training_logs', namespace='/train-model')
+def send_training_logs(msg):
+    global train_process
+    if train_process:
+        while True:
+            errors = train_process.stderr.readlines()
+            if errors:
+                for error in errors:
+                    time.sleep(0.1)
+                    emit('training_logs', {'log': error.decode().strip() + '\n'})
+                train_process.terminate()
+                train_process = None
+                break
+            else:
+                output = train_process.stdout.readline()
+                if output == b'' and train_process.poll() is not None:
+                    train_process.terminate()
+                    train_process = None
+                    break
+                if output:
+                    time.sleep(0.1)
+                    emit('training_logs', {'log': output.decode().strip() + '\n'})
+        if not errors:
+            img_path = url_for('training_bp.confusion_matrix', project=msg['project'])
+            emit('success', {'status': 'Complete', 'img_path': img_path})
+        else:
+            emit('failed', {'status': 'Failed'})
     else:
-        log = "No Training Process Running to Terminate."
-    return render_template('training.html', project=project, path=path, models=utils.BACKBONE_MODELS, is_disabled=False,
-                           logs=[log])
+        emit('status', msg)
 
 
-# @training_bp.route('/confusion-matrix/<string:project>/', methods=['GET'])
-# def get_confusion_matrix(project):
-#     img_path = os.path.join(os.getcwd(), 'dataset/', project)
-#
-#     if os.path.exists(os.path.join(img_path, 'confusion_matrix.png')):
-#         return send_from_directory(img_path, 'confusion_matrix.png', as_attachment=True)
-#     return None
+@training_bp.route('/confusion-matrix/<string:project>', methods=['GET'])
+@training_bp.route('/confusion-matrix/<string:project>/<string:output_folder>', methods=['GET'])
+def confusion_matrix(project, output_folder):
+    path = utils.lookup_project_path(project)
+    img_path = os.path.join(path, output_folder, 'checkpoints')
+    return send_from_directory(img_path, filename='confusion_matrix.png', as_attachment=True)
```

Comment:
I think this is not used

Reply:
This is being used in `training.js` where the `img` tag is being created.

Reply:
In that src will be given a path, which will render image using this method.

Reply:
Oh, I saw the `url_for` call a few lines above just now